### ISSUE TITLE
Converting russian RNC tagset to universal tagset

### ISSUE THREAD

**Issue by BLKSerene** [nltk/nltk#2151]
It seems that the tagset that nltk.pos_tag() used for russian texts is different from that used in the mapping table when nltk.map_tag() is called. So most of the tags are converted to just X.

```
>>> import nltk
>>> tokens = nltk.word_tokenize('Ру́сский язы́к ([ˈruskʲɪi̯ jɪˈzɨk] Информация о файле слушать)[~ 3][⇨] — один из восточнославянских языков, национальный язык русского народа. Является одним из наиболее распространённых языков мира — шестым среди всех языков мира по общей численности говорящих и восьмым по численности владеющих им как родным[9]. Русский является также самым распространённым славянским языком[10] и самым распространённым языком в Европе — географически и по числу носителей языка как родного[7].')
>>> nltk.pos_tag(tokens, lang = 'rus')
[('Ру́сский', 'A=m'), ('язы́к', 'S'), ('(', 'NONLEX'), ('[', 'NONLEX'), ('ˈruskʲɪi̯', 'NONLEX'), ('jɪˈzɨk', 'NONLEX'), (']', 'NONLEX'), ('Информация', 'S'), ('о', 'PR'), ('файле', 'S'), ('слушать', 'V'), (')', 'NONLEX'), ('[', 'NONLEX'), ('~', 'NONLEX'), ('3', 'NUM=ciph'), (']', 'NONLEX'), ('[', 'NONLEX'), ('⇨', 'NONLEX'), (']', 'NONLEX'), ('—', 'NONLEX'), ('один', 'A-PRO=m'), ('из', 'PR'), ('восточнославянских', 'A=pl'), ('языков', 'S'), (',', 'NONLEX'), ('национальный', 'A=m'), ('язык', 'S'), ('русского', 'A=m'), ('народа', 'S'), ('.', 'NONLEX'), ('Является', 'V'), ('одним', 'A-PRO=m'), ('из', 'PR'), ('наиболее', 'ADV'), ('распространённых', 'V'), ('языков', 'S'), ('мира', 'S'), ('—', 'NONLEX'), ('шестым', 'V'), ('среди', 'PR'), ('всех', 'A-PRO=pl'), ('языков', 'S'), ('мира', 'S'), ('по', 'PR'), ('общей', 'A=f'), ('численности', 'S'), ('говорящих', 'A=pl'), ('и', 'CONJ'), ('восьмым', 'V'), ('по', 'PR'), ('численности', 'S'), ('владеющих', 'V'), ('им', 'S-PRO'), ('как', 'CONJ'), ('родным', 'A=m'), ('[', 'NONLEX'), ('9', 'NUM=ciph'), (']', 'NONLEX'), ('.', 'NONLEX'), ('Русский', 'A=m'), ('является', 'V'), ('также', 'ADV'), ('самым', 'A-PRO=m'), ('распространённым', 'V'), ('славянским', 'A=m'), ('языком', 'S'), ('[', 'NONLEX'), ('10', 'NUM=ciph'), (']', 'NONLEX'), ('и', 'CONJ'), ('самым', 'A-PRO=m'), ('распространённым', 'A=m'), ('языком', 'S'), ('в', 'PR'), ('Европе', 'S'), ('—', 'NONLEX'), ('географически', 'ADV'), ('и', 'CONJ'), ('по', 'PR'), ('числу', 'S'), ('носителей', 'S'), ('языка', 'S'), ('как', 'CONJ'), ('родного', 'A=m'), ('[', 'NONLEX'), ('7', 'NUM=ciph'), (']', 'NONLEX'), ('.', 'NONLEX')]
>>> nltk.pos_tag(tokens, tagset = 'universal', lang = 'rus')
[('Ру́сский', 'X'), ('язы́к', 'X'), ('(', 'X'), ('[', 'X'), ('ˈruskʲɪi̯', 'X'), ('jɪˈzɨk', 'X'), (']', 'X'), ('Информация', 'X'), ('о', 'X'), ('файле', 'X'), ('слушать', 'X'), (')', 'X'), ('[', 'X'), ('~', 'X'), ('3', 'X'), (']', 'X'), ('[', 'X'), ('⇨', 'X'), (']', 'X'), ('—', 'X'), ('один', 'X'), ('из', 'X'), ('восточнославянских', 'X'), ('языков', 'X'), (',', 'X'), ('национальный', 'X'), ('язык', 'X'), ('русского', 'X'), ('народа', 'X'), ('.', 'X'), ('Является', 'X'), ('одним', 'X'), ('из', 'X'), ('наиболее', 'X'), ('распространённых', 'X'), ('языков', 'X'), ('мира', 'X'), ('—', 'X'), ('шестым', 'X'), ('среди', 'X'), ('всех', 'X'), ('языков', 'X'), ('мира', 'X'), ('по', 'X'), ('общей', 'X'), ('численности', 'X'), ('говорящих', 'X'), ('и', 'X'), ('восьмым', 'X'), ('по', 'X'), ('численности', 'X'), ('владеющих', 'X'), ('им', 'X'), ('как', 'X'), ('родным', 'X'), ('[', 'X'), ('9', 'X'), (']', 'X'), ('.', 'X'), ('Русский', 'X'), ('является', 'X'), ('также', 'X'), ('самым', 'X'), ('распространённым', 'X'), ('славянским', 'X'), ('языком', 'X'), ('[', 'X'), ('10', 'X'), (']', 'X'), ('и', 'X'), ('самым', 'X'), ('распространённым', 'X'), ('языком', 'X'), ('в', 'X'), ('Европе', 'X'), ('—', 'X'), ('географически', 'X'), ('и', 'X'), ('по', 'X'), ('числу', 'X'), ('носителей', 'X'), ('языка', 'X'), ('как', 'X'), ('родного', 'X'), ('[', 'X'), ('7', 'X'), (']', 'X'), ('.', 'X')]
```

The mapping table used in NLTK 3.3:

```
!	.
A	ADJ
AD	ADV	
C	CONJ
COMP	CONJ
IJ	X
NC	NUM
NN	NOUN
P	ADP
PTCL	PRT
V	VERB
VG	VERB
VI	VERB
VP	VERB
YES_NO_SENT	X
Z	X
```

**Comment by alvations** [comment]
The `pos_tag()` function only works on English. The `lang` argument in `pos_tag` does not load a different model, the model would still be English. 

So in that case all Russian tokens are tagged should be wrong:

```python
>>> from nltk import word_tokenize, pos_tag
>>> text = 'Ру́сский язы́к ([ˈruskʲɪi̯ jɪˈzɨk] Информация о файле слушать)[~ 3][⇨] — один из восточнославянских языков, национальный язык русского народа. Является одним из наиболее распространённых языков мира — шестым среди всех языков мира по общей численности говорящих и восьмым по численности владеющих им как родным[9]. Русский является также самым распространённым славянским языком[10] и самым распространённым языком в Европе — географически и по числу носителей языка как родного[7].'
>>> pos_tag(word_tokenize(text))
[('Ру́сский', 'NN'), ('язы́к', 'NNP'), ('(', '('), ('[', 'NNP'), ('ˈruskʲɪi̯', 'NNP'), ('jɪˈzɨk', 'NN'), (']', 'NNP'), ('Информация', 'NNP'), ('о', 'NNP'), ('файле', 'NNP'), ('слушать', 'NNP'), (')', ')'), ('[', 'VBD'), ('~', '$'), ('3', 'CD'), (']', 'NNP'), ('[', 'NNP'), ('⇨', 'NNP'), (']', 'NNP'), ('—', 'NNP'), ('один', 'NNP'), ('из', 'NNP'), ('восточнославянских', 'NNP'), ('языков', 'NNP'), (',', ','), ('национальный', 'NNP'), ('язык', 'NNP'), ('русского', 'NNP'), ('народа', 'NNP'), ('.', '.'), ('Является', 'VB'), ('одним', 'JJ'), ('из', 'NNP'), ('наиболее', 'NNP'), ('распространённых', 'NNP'), ('языков', 'NNP'), ('мира', 'NNP'), ('—', 'NNP'), ('шестым', 'NNP'), ('среди', 'NNP'), ('всех', 'NNP'), ('языков', 'NNP'), ('мира', 'NNP'), ('по', 'NNP'), ('общей', 'NNP'), ('численности', 'NNP'), ('говорящих', 'NNP'), ('и', 'NNP'), ('восьмым', 'NNP'), ('по', 'NNP'), ('численности', 'NNP'), ('владеющих', 'NNP'), ('им', 'NNP'), ('как', 'NNP'), ('родным', 'NNP'), ('[', 'VBD'), ('9', 'CD'), (']', 'NN'), ('.', '.'), ('Русский', 'CC'), ('является', 'JJ'), ('также', 'NNP'), ('самым', 'NNP'), ('распространённым', 'NNP'), ('славянским', 'NNP'), ('языком', 'NNP'), ('[', 'VBD'), ('10', 'CD'), (']', 'NNP'), ('и', 'NNP'), ('самым', 'NNP'), ('распространённым', 'NNP'), ('языком', 'NNP'), ('в', 'NNP'), ('Европе', 'NNP'), ('—', 'NNP'), ('географически', 'NNP'), ('и', 'NNP'), ('по', 'NNP'), ('числу', 'NNP'), ('носителей', 'NNP'), ('языка', 'NNP'), ('как', 'NNP'), ('родного', 'NNP'), ('[', 'VBD'), ('7', 'CD'), (']', 'NN'), ('.', '.')]

```

**Comment by BLKSerene** [comment]
There is a russian model for pos tagging. `nltk.pos_tag()` call `_get_tagger()` to initialize the perceptron tagger. If the parameter `lang` is set to `rus`, the Russian model will be loaded instead of the English one.

```
RUS_PICKLE = 'taggers/averaged_perceptron_tagger_ru/averaged_perceptron_tagger_ru.pickle'

def _get_tagger(lang=None):
    if lang == 'rus':
        tagger = PerceptronTagger(False)
        ap_russian_model_loc = 'file:' + str(find(RUS_PICKLE))
        tagger.load(ap_russian_model_loc)
    else:
        tagger = PerceptronTagger()
    return tagger
```

The PTB tagset used for English POS tagging doesn't include tags like `A=M`, `S`, `NONLEX`

**Comment by BLKSerene** [comment]
The tagset used by Russian National Corpus is listed here: http://www.ruscorpora.ru/en/corpora-morph.html

There is a pull request in the original repository `Universal Tagset` to update the mapping table: https://github.com/slavpetrov/universal-pos-tags/pull/11/files

However, since the tags sometimes have suffixes like `=m` or `=ciph`, etc. attached to them, it might be necessary to generate the mapping table to enumerate all possible tags.

**Comment by alvations** [comment]
Ah, yes https://github.com/nltk/nltk/blob/develop/nltk/tag/__init__.py#L91 was snugged in in #1784 

Hmmm... That makes it a little complicated to maintain but I guess it's okay for now if there's just English and Russian. 

----

Yes, the mappings to the universal tagger looks wrong for the Russian model added in #1784 

Even if comes from https://tech.yandex.ru/mystem/ , it's also using the new POS tagset. 

The universal pos tag mappings from https://github.com/slavpetrov/universal-pos-tags looks badly outdated =(

The pos tag conversion from UD looks different too https://github.com/UniversalDependencies/docs/blob/pages-source/_tagset-conversion

**Comment by alvations** [comment]
Thanks @BLKSerene for raising the issue! I've proposed a patch in the #2152 PR.

**Comment by alvations** [comment]
Closing issue as resolved, since #2152 is merged. 

Please re-open if the issue is still relevant/unresolved.
